param map overwrite priority

0xrofi opened the pull request:
This pull-req fix the problem that Fixed parameters can be overwritten with Get parameters.

Here is the reproduction scenario, `routes`

```
GET     /hoge            App.Hoge("fixed")
```

and controller

```go
func (c App) Hoge(param string) revel.Result {
	fmt.Println(param)
	return c.Render()
}
```

and url

```
http://localhost:9000/hoge?param=get
```

exptected output: `fixed`
but actual: `get`

Issue #1095 and pr #1126 change the order of params priority, but insufficient.
So this pr is overwrite low priority params by higher priority params.

notzippy commented:
Can you update params_test.go with a test case for this situation ? thanks..

pedromorgan commented:
Confused.. The `fixed` should overwrite all other params, but all other params need to be arrays ?

0xrofi commented:
@notzippy This situation is not only params but also include route, binder, controller. So I don't know how to write this situation's test case... sorry.

@pedromorgan Because first element of array is used to controller's arguments. So just like as I said above, `Route` parameters and `Form` parameters also can be overwritten with `Query` parameters. (Sorry for `Get` in I said above, it's a mistake of `Query`)
[problem of #1095](https://github.com/revel/revel/issues/1095#issue-173892886) isn't fixed yet and I think this patch fixes it.

If all params(`Query`, `Form`, `Route` and `Fixed`) need to append to array, I think that it should change the order of appending parameters (it's discussed in #503 ).

notzippy commented:
Would this be better ? So highest gets the first slot but lower ones can be promoted...
```
	for k, v := range p.Fixed {
-		values[k] = append(values[k], v...)
+		values[k] = append(v, values[k]...)
 	}

0xrofi commented:
Alternatively I suggest here instead of this PR.

```diff
        // Copy everything into a param map,
-       // order of priority is least to most trusted
+       // order of priority is most to least trusted
        values := make(url.Values, numParams)

-       // ?query vars first
-       for k, v := range p.Query {
+       // fixed vars first
+       for k, v := range p.Fixed {
                values[k] = append(values[k], v...)
        }
-       // form vars overwrite
-       for k, v := range p.Form {
+       // :/path vars append
+       for k, v := range p.Route {
                values[k] = append(values[k], v...)
        }
-       // :/path vars overwrite
-       for k, v := range p.Route {
+       // form vars append
+       for k, v := range p.Form {
                values[k] = append(values[k], v...)
        }
-       // fixed vars overwrite
-       for k, v := range p.Fixed {
+       // ?query vars append
+       for k, v := range p.Query {
                values[k] = append(values[k], v...)
        }
```

notzippy commented:
What about a hybrid approach ? Since fixed and route parameters should never be set using query or form values they always do a replace, and query and form parameters are appended together.
```
	// ?query vars are first
	for k, v := range p.Query {
		values[k] = append(values[k], v...)
	}
    
	// form vars append
	for k, v := range p.Form {
		values[k] = append(values[k], v...)
	}
    
	// :/path vars overwrite
	for k, v := range p.Route {
		values[k] = v 
	}
    
	// fixed vars overwrite
	for k, v := range p.Fixed {
		values[k] = v 
	}
```

pedromorgan commented:
@notzippy Yes that is what we want.. All query/post etc vars appended.. BUT a fixed one is CONST ;-) ta.. commit it bro...

pedromorgan commented:
Need to update manual to explain this problem.. and also a new CORS chapter.. ;-)

notzippy commented:
@pedromorgan I think with this PR #1157 you may need to rewrite the manual :-P
I incorporated this change in there..

brendensoares commented:
Closing in favor of the above defined hybrid approach to be released with PR #1171. Thanks for everyone's effort on this!
